### ISSUE TITLE
Update Notification Title during playback

### ISSUE THREAD

**Issue by AmrSubZero** [androidx/media#951]
Using **Media3 ExoPlayer@1.2.0** with a ``PlaybackService`` that extends **MediaSessionService**

using ```setMediaNotificationProvider``` to create a ```Notification``` for API below 33.

```java
public class PlaybackService extends MediaSessionService {

    public void onCreate() {

        // Initialize ExoPlayer
        ExoPlayer player = new ExoPlayer.Builder(this, ...).build();

        setMediaNotificationProvider(new MediaNotification.Provider() {
            @NonNull
            @Override
            public MediaNotification createNotification(@NonNull MediaSession mediaSession, @NonNull ImmutableList<CommandButton> customLayout, @NonNull MediaNotification.ActionFactory actionFactory, @NonNull Callback onNotificationChangedCallback) {

                createMediaNotification(mediaSession);

                return new MediaNotification(PLAYBACK_NOTIFICATION_ID, notificationBuilder.build());
            }

            // ...
        });

        // ...

    });

    public void createMediaNotification(MediaSession mediaSession) {

        MediaMetadata metadata = mediaSession.getPlayer().getMediaMetadata();

        NotificationCompat.Builder notificationBuilder = new NotificationCompat.Builder(PlaybackService.this, PLAYBACK_CHANNEL_ID)
            // ...
            .setContentTitle(metadata.title)
            .setContentText(metadata.description)
            .setStyle(new MediaStyleNotificationHelper.MediaStyle(mediaSession));

    }

}
```

in ```MainActivity``` i set the ```MediaMetaData``` (title/description)

```java
List<MediaItem> mediaItems = new ArrayList<>();

MediaItem mediaItem = new MediaItem.Builder()
    .setMediaId(ITEM_ID)
    .setMediaMetadata(new MediaMetadata.Builder()
            .setTitle(ITEM_TITLE)
            .setDescription(ITEM_DESCRIPTION)
            // ..
    .setUri(ITEM_URI)
.build();

mediaItems.add(mediaItem);

player.setMediaItems(mediaItems);

player.prepare();
player.play();
```

**TLDR:** i need to update title/description of the media playback notification (during playback), after it's already set before using ```.setMediaMetadata(new MediaMetadata.Builder().setTitle(..))``` and ```.setContentTitle(metadata.title)```

Now, what i'm trying to achieve, lets say at some point ```(during playback)``` i need to **update the title/description** (or MediaMetaData title/description) of the Notification for the current playing media item, wether it is the ```Media3 Session Notification``` or ```Legacy Notification``` below API Level 33

```java
// a method in MainActivity
private void onPlayerProgressChanged(long current_playback_time) {

    if(current_playback_time >= something) {
        
        // Here need to update the notification title/description
        // Like .setDescription("Something else")
    }

}
```

Here's how it's currently working:

![Notifications](https://github.com/androidx/media/assets/5399778/09c89493-cd73-4894-abac-f56467c48cc2)

I can't seem to find any documentation or workaround to achieve this approach.

How can i achieve this?

**Comment by tianyif** [comment]
Hi @AmrSubZero,

You can utilize the [`replaceMediaItem`](https://github.com/androidx/media/blob/release/libraries/exoplayer/src/main/java/androidx/media3/exoplayer/ExoPlayer.java#L1504) to achieve this. 

```
MediaItem oldItem = ... // Your original item
MediaItem newItem = oldItem
                        .buildUpon()
                        .setTitle("new title")
                        .setDescription("new description")
                        .build();
player.replaceMediaItem(itemIndex, newItem);
```

In this way, the player can emit an `EVENT_MEDIA_METADATA_CHANGED` event, and media3 session [`MediaNotificationManager`](https://github.com/androidx/media/blob/release/libraries/session/src/main/java/androidx/media3/session/MediaNotificationManager.java#L357) or legacy [`PlayerNotificationManager`](https://github.com/androidx/media/blob/release/libraries/ui/src/main/java/androidx/media3/ui/PlayerNotificationManager.java#L1549) can update the notification when receive the metadata changed event.

Hope this helps!

**Comment by AmrSubZero** [comment]
Hello @tianyif

Actually the ```player.replaceMediaItem()``` did work ```updating the title```, but i think it does not suit my case, here's why

The logic, since **Media3** or **ExoPlayer** does not have built in ```Position``` tracker/listener *(as far as i know)*, so i built my own using (Runnable/Handler) runs every 1 second.

<details>
  <summary>See here</summary>
  
```java
// ProgressTracker is a custom class that tracks player.getCurrentPosition() every 1 second.
ProgressTracker tracker;

MediaMetadata metadata = new MediaMetadata.Builder()....build();

player.addListener(new Player.Listener() {

    @Override
    public void onIsPlayingChanged(boolean isPlaying) {

        if(isPlaying) {

            tracker = new ProgressTracker(player, new ProgressTracker.PositionListener() {

                @Override
                public void progress(long current_position) {

                    if(current_position >= something) {

                        // Replace the current MediaItem with new Title
                        MediaItem mediaItem = player.getCurrentMediaItem();
                        if(mediaItem != null) {
                            mediaItem
                                .buildUpon()
                                .setMediaMetadata(metadata.buildUpon()
                                        .setTitle("Part: x")
                                        .build())
                                .build();
                            player.replaceMediaItem(0, mediaItem);
                        }
                    }
                }

            }

        }else {
            // Purge the position tracker (handler.removeCallbacks())
            if(tracker != null) { tracker.purgeHandler(); }
        }

    }

}


@Override
public void onDestroy() {

    // Purge the position tracker (handler.removeCallbacks())
    if(tracker != null) { tracker.purgeHandler(); }
}
```
</details>

Then i check if the current position equals something, there i need to update the meta title like **"Part: 13"**, this runs very frequently!

Using ```player.replaceMediaItem()``` kinda resets the current position of the player for the second it did a replace, then it gets back to normal position.

<details>
  <summary>See here</summary>
  
```sql
position: 50		--> first
position: 845
position: 1846
position: 2849
position: 3858
-- called player.replaceMediaItem()
position: 49		-- > should be 4xxx
position: 1051		-- > should be 5xxx
position: 6890
position: 7892
-- called player.replaceMediaItem()
position: 118		-- > should be 8xxx
position: 9994
position: 10996
position: 11999
position: 12993
position: 13994
-- called player.replaceMediaItem()
position: 115		-- > should be 14xxx
position: 16030
position: 17031
position: 18033
-- called player.replaceMediaItem()
position: 124		-- > should be 19xxx
position: 1124		-- > should be 20xxx
position: 2125		-- > should be 21xxx
position: 22074
-- called player.replaceMediaItem()
position: 101		-- > should be 23xxx
position: 1103		-- > should be 24xxx
position: 25080
position: 26080
position: 27081
-- called player.replaceMediaItem()
position: 100		-- > should be 28xxx
position: 1101		-- > should be 29xxx
position: 2102		-- > should be 30xxx
position: 31124
position: 32125
-- called player.replaceMediaItem()
position: 75		-- > should be 33xxx
position: 34126
position: 35128
```
</details>

Therefore, when i get incorrect ```current position``` things go wrong.

also is it a best practice calling ```player.replaceMediaItem()``` for a time like every 5 seconds? i can think it will break someway somehow!

So, my hopes:

- another way to update the current playback (notification/mediasession) ```title``` without needing to replace the current media item.
- a proper way to track current playback position (maybe built-in?) that doesn't get affected with things like ```player.replaceMediaItem()```

Thanks!

**Comment by AmrSubZero** [comment]
Hello @tianyif thanks for keeping up!

after hours of debugging and trying several methods, here's what i've discovered!

- We have ```Exoplayer player``` inside the ```MediaSessionService```

```java
// PlaybackService.java

public class PlaybackService extends MediaSessionService {

    ExoPlayer player;

    @Override
    @UnstableApi
    public void onCreate() {

        player = new ExoPlayer.Builder(this).build();
    }
}
```

- and we have ```MediaController player``` in ```MainActivity``` which i receive as documented

```java
// MainActivity.java

MediaController player;

ListenableFuture<MediaController> controllerFuture;

@Override
protected void onStart() {

    SessionToken sessionToken =
            new SessionToken(this, new ComponentName(this, PlaybackService.class));
    controllerFuture =
            new MediaController.Builder(this, sessionToken).buildAsync();
    controllerFuture.addListener(() -> {

        try {

            player = controllerFuture.get();

        } catch (Exception e) { e.printStackTrace(); }

    }, MoreExecutors.directExecutor());

}
```

When i tested getting ```current position``` from both *(with different methods)* as we discussed, with doing ```player.replaceMediaItem()``` frequently *(every 1 second)* :

- **MediaSessionService** ```ExoPlayer player``` always giving the correct ```current position``` and it never gets interrupted by calling ```player.replaceMediaItem()```

- **MainActivity** ```MediaController player``` somehow gets interrupted by calling ```player.replaceMediaItem()``` and giving incorrect *(position reset)* right after calling the **.replaceMediaItem**

i tried several methods to track ```current position``` here's by getting it directly from ```PlayerView```

<details>
  <summary>See here</summary>

```java
public class MainActivity extends AppCompatActivity {

    MediaController player;

    ListenableFuture<MediaController> controllerFuture;

    PlayerView playerView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        playerView = findViewById(R.id.playerView);
    }

    private void getCurrentPlayerPosition() {

        Log.i("PLAYBACK_STATE", "Playback Position: " + player.getCurrentPosition());

        if (player.isPlaying()) {
            playerView.postDelayed(this::getCurrentPlayerPosition, 1000);
        }
    }

    @Override
    protected void onStart() {

        SessionToken sessionToken =
                new SessionToken(this, new ComponentName(this, PlaybackService.class));
        controllerFuture =
                new MediaController.Builder(this, sessionToken).buildAsync();
        controllerFuture.addListener(() -> {

            try {

                player = controllerFuture.get();

                // Set Player to the PlayerView
                playerView.setPlayer(player);

                player.addListener(new Player.Listener() {

                    @Override
                    public void onIsPlayingChanged(boolean isPlaying) {

                        if(isPlaying) {
                            // Track the current playback position (every 1 second)
                            playerView.postDelayed(MainActivity.this::getCurrentPlayerPosition, 1000);
                        }
                }

            } catch (Exception e) { e.printStackTrace(); }

        }, MoreExecutors.directExecutor());

    }

}
```

</details>

Either way, updaing the **MediaMetaData** title *(frequently)* is working now by calling ```.replaceMediaItem()``` thanks to your support.

But i'm still curious, why the ```current poisition``` gets interrupted from ```MediaController player``` in ```MainActivity``` and it does not from ```ExoPlayer player``` in ```MediaSessionService```

Thanks! 👏

**Comment by tianyif** [comment]
Hi @AmrSubZero,

Thanks for the impressive details! Yes, I could reproduce the interrupted position with the [test controller app](https://github.com/googlesamples/android-media-controller/blob/master/mediacontroller/src/main/java/com/example/android/mediacontroller/MediaAppControllerActivity.java) with connecting to a session from the demo session app.

The reason being this is based on the three facts - 

1. When the `MediaItem` is replaced at the `MediaController`, the `PlayerInfo` (including the position info) got [masked](https://github.com/androidx/media/blob/release/libraries/session/src/main/java/androidx/media3/session/MediaControllerImplBase.java#L1141) by at the controller side, with the default position of the new item (default is 0). 
2. Meanwhile, it is expected to receive the player info update from the session triggered by `mediaMetadata` change, after the `MediaItem` is replaced at the session player. 
3. The session side is regularly sending the `SessionPositionInfo` to the controller (default interval is 3s), and controller will then update its `playerInfo.SessionPositionInfo`.
4. When `getCurrentPosition()` at the controller side, the returned current position is [interpolated](https://github.com/androidx/media/blob/release/libraries/session/src/main/java/androidx/media3/session/MediaUtils.java#L1541-L1543) by the latest `playerInfo.SessionPositionInfo.positionInfo.positionMs` and the elapsed time from the last update.

Then I added some logs and found that -

```
2024-01-11 14:31:01.172 24535-24535 PLAYBACK_STATE          androidx.media3.testapp.controller   W  [1] Replace MediaItem, masked position: 0
2024-01-11 14:31:01.373 24535-24535 PLAYBACK_STATE          androidx.media3.testapp.controller   W  [2] onPlayerInfoUpdated update position to: 81
2024-01-11 14:31:02.131 24535-24535 PLAYBACK_STATE          androidx.media3.testapp.controller   W  [2] onPlayerInfoUpdated update position to: 980
2024-01-11 14:31:02.179 24535-24535 PLAYBACK_STATE          androidx.media3.testapp.controller   W  [4] controller latest playerInfo.SessionPositionInfo.positionInfo.positionMs: : 980
2024-01-11 14:31:02.179 24535-24535 PLAYBACK_STATE          androidx.media3.testapp.controller   I  position: 1043
2024-01-11 14:31:02.181 24535-24535 PLAYBACK_STATE          androidx.media3.testapp.controller   W  [1] Replace MediaItem, masked position: 0
2024-01-11 14:31:02.323 24535-24535 PLAYBACK_STATE          androidx.media3.testapp.controller   W  [2] onPlayerInfoUpdated update position to: 1038
2024-01-11 14:31:03.182 24535-24535 PLAYBACK_STATE          androidx.media3.testapp.controller   W  [4] controller latest playerInfo.SessionPositionInfo.positionInfo.positionMs: : 1038
2024-01-11 14:31:03.182 24535-24535 PLAYBACK_STATE          androidx.media3.testapp.controller   I  position: 1919
2024-01-11 14:31:03.185 24535-24535 PLAYBACK_STATE          androidx.media3.testapp.controller   W  [1] Replace MediaItem, masked position: 0
2024-01-11 14:31:04.187 24535-24535 PLAYBACK_STATE          androidx.media3.testapp.controller   W  [4] controller latest playerInfo.SessionPositionInfo.positionInfo.positionMs: : 0
2024-01-11 14:31:04.187 24535-24535 PLAYBACK_STATE          androidx.media3.testapp.controller   W  WRONG position estimate: 1003
2024-01-11 14:31:04.187 24535-24535 PLAYBACK_STATE          androidx.media3.testapp.controller   I  position: 1003
2024-01-11 14:31:04.190 24535-24535 PLAYBACK_STATE          androidx.media3.testapp.controller   W  [1] Replace MediaItem, masked position: 0
2024-01-11 14:31:05.124 24535-24535 PLAYBACK_STATE          androidx.media3.testapp.controller   W  [3] Periodically update SessionPositionInfo from position: 0 , to: 3980
2024-01-11 14:31:05.191 24535-24535 PLAYBACK_STATE          androidx.media3.testapp.controller   W  [4] controller latest playerInfo.SessionPositionInfo.positionInfo.positionMs: : 3980
2024-01-11 14:31:05.191 24535-24535 PLAYBACK_STATE          androidx.media3.testapp.controller   I  position: 4053
2024-01-11 14:31:05.194 24535-24535 PLAYBACK_STATE          androidx.media3.testapp.controller   W  [1] Replace MediaItem, masked position: 0
2024-01-11 14:31:05.285 24535-24535 PLAYBACK_STATE          androidx.media3.testapp.controller   W  [2] onPlayerInfoUpdated update position to: 4116
2024-01-11 14:31:06.195 24535-24535 PLAYBACK_STATE          androidx.media3.testapp.controller   W  [4] controller latest playerInfo.SessionPositionInfo.positionInfo.positionMs: : 4116
2024-01-11 14:31:06.195 24535-24535 PLAYBACK_STATE          androidx.media3.testapp.controller   I  position: 5049
2024-01-11 14:31:06.199 24535-24535 PLAYBACK_STATE          androidx.media3.testapp.controller   W  [1] Replace MediaItem, masked position: 0
2024-01-11 14:31:06.235 24535-24535 PLAYBACK_STATE          androidx.media3.testapp.controller   W  [2] onPlayerInfoUpdated update position to: 5055
2024-01-11 14:31:07.199 24535-24535 PLAYBACK_STATE          androidx.media3.testapp.controller   W  [4] controller latest playerInfo.SessionPositionInfo.positionInfo.positionMs: : 5055
2024-01-11 14:31:07.199 24535-24535 PLAYBACK_STATE          androidx.media3.testapp.controller   I  position: 6027
```

The wrong position was marked as WRONG, which is 1003. I marked the logs corresponding to each above steps, and you can see between replacing the item and getting the wrong position 1003, there is neither [2] and [3] that update the `SessionPositionInfo`. The elapsed time since the last update is 1003, so plus masked position which is 0, we wrongly got 1003.

My rough thought is around whether we should use the default position 0 to mask the current position when the item is being replaced (corresponding to the above [1]). It makes sense in the case that the current playing item is completely removed and the default position of the next item will be immediately used, but in the case that the current playing item is replaced, using the current position instead of 0 would be a better option.

We will further look into and provide a proper solution to it. Thanks a lot for all the details and repo of the project!
